Tiktok pixel integration

jae-rhee-tiktok opened the pull request:
TikTok Pixel Browser Destination Action

silesky commented:
@jae-rhee-tiktok Hey, this is in the wrong folder.

silesky commented on packages/browser-destinations/src/destinations/tiktok-pixel/reportWebEvent/formatter.ts:
```diff
@@ -0,0 +1,46 @@
+import { createHash } from 'crypto'
```
This isn't available in the browser?

pooyaj reviewed (changes requested):
We need to the structure of the destination a little bit due to some recent structural changes in the repo ( sorry for the hassle ): 
- instead of `packages/browser-destinations/src/destinations` should be `packages/browser-destinations/destinations`
- We need to restructure of the code as well, the code should be moved to `/src` directory, and we need to add a `package.json` and `tsconfig.json` to the destination's root directory ( please check other destinations to see examples )

jae-rhee-tiktok commented on packages/browser-destinations/src/destinations/tiktok-pixel/index.ts:
```diff
@@ -0,0 +1,150 @@
+// Switch from unknown to the partner SDK client types
+export const destination: BrowserDestinationDefinition<Settings, TikTokPixel> = {
+  name: 'TikTok Pixel',
+  slug: 'tiktok-pixel-actions',
```
Thank you - updated the slug name to `actions-tiktok-pixel`

jae-rhee-tiktok commented:
Hi @pooyaj, I updated the structure and added the package.json and tsconfig.json files to the tiktok pixel directory. Also, I removed to crypto function which is not available. Thank you!

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,145 @@
+})
+
+describe('ttq.track', () => {
+  test('it maps the event name and properties and passes them into RipeSDK.track', async () => {
```
```suggestion
  test('it maps the event name and properties and passes them into TikTokDestination.reportWebEvent', async () => {
```

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,145 @@
+      })
+    )
+
+    expect(destination.actions.track.perform).toHaveBeenCalledWith(
```
This probably should be like this: 
```suggestion
    expect(destination.actions.reportWebEvent.perform).toHaveBeenCalledWith(
```

pooyaj reviewed (commented):
Looks like tests are failing. Some of the paths in the tests are incorrect. Also added few comments on the test logic as well. Let's update and see if we can get the tests passing locally. 

Also may worth testing different scenarios for all the presets available for this destination.

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,370 @@
+    const [event] = await TikTokDestination({
+      subscriptions,
+      pixelCode: '123',
+      useExistingPixel: false
```
we probably need to set this to `true` in order to use the mocked `window.ttq` value. Otherwise, the tests run the `initScript` which leads to exceptions for some reason.

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,370 @@
+
+    const ajs = new Analytics({ writeKey: '123' })
+    await event.load(Context.system(), ajs)
+    jest.spyOn(destination.actions.reportWebEvent.perform, 'perform')
```
```suggestion
    jest.spyOn(destination.actions.reportWebEvent, 'perform')
```

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,370 @@
+
+    const ajs = new Analytics({ writeKey: '123' })
+    await event.load(Context.system(), ajs)
+    jest.spyOn(destination.actions.reportWebEvent.perform, 'perform')
```
```suggestion
    jest.spyOn(destination.actions.reportWebEvent, 'perform')
```

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,370 @@
+
+    const ajs = new Analytics({ writeKey: '123' })
+    await event.load(Context.system(), ajs)
+    jest.spyOn(destination.actions.reportWebEvent.perform, 'perform')
```
```suggestion
    jest.spyOn(destination.actions.reportWebEvent, 'perform')
```

pooyaj reviewed (commented):
- Let's run the tests using the following command from the repo's root directory: 
```
yarn browser test /Users/pooyajaferian/dev/src/github.com/segmentio/action-destinations/packages/browser-destinations/destinations/tiktok-pixel
```
- Made few suggestions to fix some errors that I saw when running tests locally
- Also, a more through approach to test would be to check how the TikTok pixel is being called rather than testing how the `perform` block is being called. Here is an example here where we mock Hubspot SDK, and verify its being called with the right set of parameters: https://github.com/segmentio/action-destinations/blob/main/packages/browser-destinations/destinations/hubspot-web/src/trackCustomBehavioralEvent/__tests__/index.test.ts#L86-L100

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,136 @@
+          groupId: {
+            '@path': '$.groupId'
+          },
+          event: {
+            '@path': 'PlaceAnOrder'
+          },
```
```suggestion
          event: 'PlaceAnOrder',
```

pooyaj commented on packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts:
```diff
@@ -0,0 +1,136 @@
+    await reportWebEvent.load(Context.system(), {} as Analytics)
+    await reportWebEvent.track?.(context)
+
+    expect(destination.actions.reportWebEvent.perform).toHaveBeenCalledWith([
+      ['page'],
```
So this statement checks what the `perform` block of our destination will be called with: 
- First param is `ttq` ( that's why we are seeing array of `page`/`identify`/`track`
- Second param is an object which will be passed in by actions framework, containing payload, context, etc. 

Ideally we want to mock `ttq` and then assert on something like this: 
```
expect(mockTtq.page).toHaveBeenCalledWith ...
expect(mockTtq.identify).toHaveBeenCalledWith ...
expect(mockTtq.track).toHaveBeenCalledWith ...
```

jae-rhee-tiktok commented:
Hi Pooya,

Thank you for the clarification. I was able to run the test cases
successfully and added new ones for other preset types:

   1. Track event with multiple products ("PlaceAnOrder")
   2. Track event with single product ("AddToCart")
   3. Page event with single product ("ViewContent")



Thank you so much!

Best,
Jae

*Jaehyuk Rhee*
Monetization Technology at TikTok

5800 Bristol Parkway, Suite 100,
Culver City, CA 90230
Tel: (401) 408-0504
From: "Pooya ***@***.***>
Date: Thu, Jul 20, 2023, 16:59
Subject: [External] Re: [segmentio/action-destinations] Tiktok pixel
integration (PR #1418)
To: ***@***.***>
Cc: "Jae ***@***.***>, "Mention"<
***@***.***>

***@***.**** commented on this pull request.
------------------------------

In
packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts
<https://github.com/segmentio/action-destinations/pull/1418#discussion_r1269930050>
:

> +          event: {
+            ***@***.***': 'PlaceAnOrder'
+          },

⬇️ Suggested change

-          event: {
-            ***@***.***': 'PlaceAnOrder'
-          },
+          event: 'PlaceAnOrder',

------------------------------

In
packages/browser-destinations/destinations/tiktok-pixel/src/reportWebEvent/__tests__/index.test.ts
<https://github.com/segmentio/action-destinations/pull/1418#discussion_r1269957103>
:

> +    expect(destination.actions.reportWebEvent.perform).toHaveBeenCalledWith([
+      ['page'],

So this statement checks what the perform block of our destination will be
called with:

   - First param is ttq ( that's why we are seeing array of page/identify/
   track
   - Second param is an object which will be passed in by actions
   framework, containing payload, context, etc.

Ideally we want to mock ttq and then assert on something like this:

expect(mockTtq.page).toHaveBeenCalledWith ...
expect(mockTtq.identify).toHaveBeenCalledWith ...
expect(mockTtq.track).toHaveBeenCalledWith ...

—
Reply to this email directly, view it on GitHub
<https://github.com/segmentio/action-destinations/pull/1418#pullrequestreview-1540028747>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/A6TWJNATHHW3NAA4SDBTFKDXRGL3FANCNFSM6AAAAAA2O4QRJA>
.
You are receiving this because you were mentioned.Message ID:
***@***.***>
